README.md
# Muta benchmark

## Requirement

- NodeJS >= 10

## Quick start

```shell
git clone https://github.com/nervosnetwork/muta-benchmark.git
cd muta-benchmark
npm install
node index -d 5 http://127.0.0.1:8000/graphql
```

or install globally

```
git clone https://github.com/nervosnetwork/muta-benchmark.git
cd muta-benchmark
npm install
[sudo] npm link
muta-bench -d 5 http://127.0.0.1:8000/graphql
```

## Options

```
Options:
  -m --method [method]            HTTP method (default: "POST")
  -d --duration [duration]        number of second (default: 5)
  -g --gap [gap]                  muta's timeout_gap (default: 20)
  -c --connections [connections]  number of connection (default: 20)
  --pk [privateKey]               sender of the transfer
  --receiver [receiver]           receiver of the transfer
  --verbose                       show verbose info, use it for debug (default: false)
  -h, --help                      output usage information
```

## Q&A

### What can I do when unknown error appeared?

Try to use `--verbose` for more bench information

### How to customize private key of the sender?

- use environment variable `PRIVATE_KEY=0x...`
- or use `--pk 0x...` option

### How to run with multiple nodes

```
# each node will be allocated 15 connections
muta-bench -c 30 http://127.0.0.1:8000/graphql http://127.0.0.1:8001/graphql
```

It is recommended to adjust the number of connections to n times the number of nodes


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.2.0](https://github.com/nervosnetwork/muta-benchmark/compare/v0.1.4...v0.2.0) (2020-09-15)


### ⚠ BREAKING CHANGES

* update to compatible with nervosnetwork/muta#320

### Features

* support multiple signatures with  nervosnetwork/muta[#318](https://github.com/nervosnetwork/muta-benchmark/issues/318) ([b4d4473](https://github.com/nervosnetwork/muta-benchmark/commit/b4d4473d1bf7f8fb986f0dcb418079289d890a51))
* supported bench with multiple url ([4e6c444](https://github.com/nervosnetwork/muta-benchmark/commit/4e6c4448bdfe84fc9fc1031fb9b2789ac0a7da81))
* update muta-sdk version to fit prevHash ([6df841a](https://github.com/nervosnetwork/muta-benchmark/commit/6df841a84fee47521839d1a37b4d5554c03d09ba))
* update sdk to 0.10 ([265bc2a](https://github.com/nervosnetwork/muta-benchmark/commit/265bc2a5af9e2dd63fe8ade4efd5a3e64070007f))
* update sdk to 0.2.0-alpha.1 ([0de7039](https://github.com/nervosnetwork/muta-benchmark/commit/0de7039259d26c296b2d3b7b70b238c9fb5b98a0))
* update sdk version ([0563fee](https://github.com/nervosnetwork/muta-benchmark/commit/0563feee02b5e2c6cb4334d0caeefedbdcfb233a))
* update to compatible with nervosnetwork/muta[#320](https://github.com/nervosnetwork/muta-benchmark/issues/320) ([9f14a6c](https://github.com/nervosnetwork/muta-benchmark/commit/9f14a6c7743ad178b159237668f1f365c5515d82))
* update to fit to bech32 address ([bd976d0](https://github.com/nervosnetwork/muta-benchmark/commit/bd976d088f8973fec0d923ce96db9582fcceb46f))

### [0.1.12](https://github.com/nervosnetwork/muta-benchmark/compare/v0.1.11...v0.1.12) (2020-08-03)


### Features

* update sdk to 0.2.0-alpha.1 ([0a91bd3](https://github.com/nervosnetwork/muta-benchmark/commit/0a91bd3875100b09c9c071d86854d7d2cad494d0))

### [0.1.11](https://github.com/nervosnetwork/muta-benchmark/compare/v0.1.11-pr374.0...v0.1.11) (2020-07-30)

### [0.1.11-pr374.0](https://github.com/nervosnetwork/muta-benchmark/compare/v0.1.10...v0.1.11-pr374.0) (2020-07-27)


### Features

* update to fit to bech32 address ([13fda98](https://github.com/nervosnetwork/muta-benchmark/commit/13fda98d62485a5c34a6042d583c262ad4f3b62c))

### [0.1.10](https://github.com/nervosnetwork/muta-benchmark/compare/v0.1.9...v0.1.10) (2020-07-13)


### Features

* update sdk version ([04b1988](https://github.com/nervosnetwork/muta-benchmark/commit/04b1988721766c1f971e0ac8cda00fbc0be36e09))

### [0.1.9](https://github.com/nervosnetwork/muta-benchmark/compare/v0.1.8...v0.1.9) (2020-06-25)


### ⚠ BREAKING CHANGES

* update to compatible with nervosnetwork/muta#320

### Features

* update to compatible with nervosnetwork/muta[#320](https://github.com/nervosnetwork/muta-benchmark/issues/320) ([c1fa8dd](https://github.com/nervosnetwork/muta-benchmark/commit/c1fa8ddf3d22da49cdb6a959586b9eab85b7efb7))

### [0.1.8](https://github.com/nervosnetwork/muta-benchmark/compare/v0.1.7...v0.1.8) (2020-06-22)


### Features

* support multiple signatures with  nervosnetwork/muta[#318](https://github.com/nervosnetwork/muta-benchmark/issues/318) ([7c0f569](https://github.com/nervosnetwork/muta-benchmark/commit/7c0f569a345028ade1a5f73a8b570f3e485a9f11))

### [0.1.7](https://github.com/homura/muta-benchmark/compare/v0.1.6...v0.1.7) (2020-05-29)


### Features

* update muta-sdk version to fit prevHash ([32176be](https://github.com/homura/muta-benchmark/commit/32176bea2703243b64e8cf95804cd52dac4cbabe))

### [0.1.6](https://github.com/homura/muta-benchmark/compare/v0.1.5...v0.1.6) (2020-05-19)


### Features

* supported bench with multiple url ([9210cf1](https://github.com/homura/muta-benchmark/commit/9210cf15909bd4b989978b9960b3bc9c4d0e33a6))

### [0.1.5](https://github.com/homura/muta-benchmark/compare/v0.1.4...v0.1.5) (2020-04-04)


### Features

* update sdk to 0.10 ([291e838](https://github.com/homura/muta-benchmark/commit/291e8380d880b2d94eba355373ef51c277a8c5cd))

### [0.1.4](https://github.com/homura/muta-benchmark/compare/v0.1.3...v0.1.4) (2020-03-10)


### Bug Fixes

* compatible with millisecond and second timestamp ([71bdabd](https://github.com/homura/muta-benchmark/commit/71bdabd33fdd9ad653a1b1c2132b3c1e0ad6b1d0))

### [0.1.3](https://github.com/homura/muta-benchmark/compare/v0.1.2...v0.1.3) (2020-03-10)


### Features

* force hex string to start with 0x ([c915ab6](https://github.com/homura/muta-benchmark/commit/c915ab6e5aadee9653eea81e05c97e92c461d4fa))
* supported [millis timestamp](https://github.com/nervosnetwork/muta/pull/216) ([ea3b25f](https://github.com/homura/muta-benchmark/commit/ea3b25f655cc00779d097a7e53bacc52b4e52724))

### [0.1.2](https://github.com/homura/muta-benchmark/compare/v0.1.1...v0.1.2) (2020-02-26)


### Bug Fixes

* fix failure caused by custom `gap` or `chain-id` ([c959b25](https://github.com/homura/muta-benchmark/commit/c959b254ed58555daf2e1d5ddfbc9b7fd9bccafe))

### [0.1.1](https://github.com/homura/muta-benchmark/compare/v0.1.0...v0.1.1) (2020-02-23)


### Features

* add json output ([#5](https://github.com/homura/muta-benchmark/issues/5)) ([a2ac419](https://github.com/homura/muta-benchmark/commit/a2ac4190a55d703af8f4535f6e3c6bc4ede4958c))
* pretty json output ([e3a71ce](https://github.com/homura/muta-benchmark/commit/e3a71ceaf3cf03babe42e7983a61deed23c776c8))

## [0.1.0](https://github.com/homura/muta-benchmark/compare/v0.0.2...v0.1.0) (2020-02-22)


### Bug Fixes

* Add Content-type to http.headers ([#4](https://github.com/homura/muta-benchmark/issues/4)) ([5a8a911](https://github.com/homura/muta-benchmark/commit/5a8a9118bc45554464a21cfc651789150faf63bc))

### 0.0.2 (2020-02-14)


### Features

* output round usage when TPS ([aea7c53](https://github.com/homura/muta-benchmark/commit/aea7c53f677ddc4ed2356e8edda56d614d13e0e1))

index.js
#!/usr/bin/env node

const program = require('commander');
const ora = require('ora');
const cluster = require('cluster');
const { utils, Account } = require('@mutadev/muta-sdk');
const { runMain, runWorker } = require('./bench');
const logger = require('./logger');
const { AssetBenchProducer } = require('./BenchProducer');

const toNumber = (v) => parseInt(v);

const receiverAccount = new Account(
  '0x0000000000000000000000000000000000000000000000000000000000000001'
);

const { args } = program
  .option('-m --method [method]', 'HTTP method', 'POST')
  .option('-d --duration [duration]', 'number of second', toNumber, 60)
  .option('-g --gap [gap]', "muta's timeout_gap", toNumber, 20)
  .option(
    '-c --connections [connections]',
    'number of connection',
    toNumber,
    20
  )
  .option(
    '--pk [privateKey]',
    'sender of the transfer',
    utils.toHex,
    process.env.PRIVATE_KEY ||
      '0x45c56be699dca666191ad3446897e0f480da234da896270202514a0e1a587c3f'
  )
  .option(
    '--chain-id [chainId]',
    'chain id',
    utils.toHex,
    '0xb6a4d7da21443f5e816e8700eea87610e6d769657d6b8ec73028457bf2ca4036'
  )
  .option(
    '--receiver [receiver]',
    'receiver of the transfer',
    utils.toHex,
    receiverAccount.address
  )
  .option('--verbose', 'show verbose info, use it for debug', false)
  .option('--cpu [cpu]', 'cpu nums', toNumber, 3)
  .option('--json', 'print output as json', false)
  .name('muta-bench')
  .usage(
    '-m POST -d 60 -c 20 --gap 30 --pk 0x45c56be699dca666191ad3446897e0f480da234da896270202514a0e1a587c3f http://127.0.0.1:8000/graphql'
  )
  .parse(process.argv);

const opts = program.opts();
const url = args.length ? args : 'http://127.0.0.1:8000/graphql';

opts.url = url;

if (opts.verbose) {
  logger.enable('*');
}

if (cluster.isMaster) {
  run().catch(({ errorCount }) => {
    if (errorCount > 0) {
      console.error(`HTTP success but GraphQL error count: ${errorCount}`);
    }
  });
} else {
  runWorker();
}

async function run() {
  console.log(opts);
  const { gap, pk, url, receiver, chainId } = opts;
  const assetBenchProducer = new AssetBenchProducer({
    pk,
    chainId,
    gap,
    url,
    receiver,
  });

  const createAssetSpin = ora('Creating asset').start();
  try {
    const asset = await assetBenchProducer.createAsset();
    createAssetSpin.succeed(`Created asset ${JSON.stringify(asset)}`);
  } catch (e) {
    createAssetSpin.fail(`Asset create failed, ${e.message}`);
  }

  await assetBenchProducer.prepare();

  opts.headers = {};
  opts.headers['Content-Type'] = 'application/json';

  const workers = Array.from({ length: opts.cpu }).map(() => {
    const worker = cluster.fork({
      WORKER_DATA: JSON.stringify({
        assetId: assetBenchProducer.assetId,
        to: assetBenchProducer.to,
        timeout: assetBenchProducer.timeout,
        chainId: assetBenchProducer.chainId,
        privateKey: utils.rm0x(utils.toHex(assetBenchProducer.privateKey)),
        sender: assetBenchProducer.account.address,
      }),
      OPTIONS: JSON.stringify(opts),
    });

    worker.on('exit', () => console.log('worker done'));
    worker.on('error', (err) => console.log('worker error', err));

    return worker;
  });

  return runMain(assetBenchProducer, workers, opts);
}
